### ISSUE TITLE
Unhandled exception occurs when enabling Truffle DB with existing artifacts

### ISSUE THREAD

**Issue by gnidan** [trufflesuite/truffle#4614]
## Issue

If you turn on Truffle DB in a project that has existing artifacts, an exception occurs when you run `truffle migrate`

## Steps to Reproduce

1. `truffle init`
2. `truffle develop` and `migrate` inside that session
3. Quit `truffle develop`
4. Enable Truffle DB
5. `truffle develop` then `migrate` again
6. Observe exception

## Expected Behavior

Data should save to Truffle DB and no error should appear

## Actual Results

An error appears. I haven't checked if the data gets saved or not, but I'm guessing it's incomplete.

## Environment

* Operating System: 
* Ethereum client:
* Truffle version (`truffle version`): v5.4.28
* node version (`node --version`):
* npm version (`npm --version`):

**Comment by tenthirtyone** [comment]
This occurs because `project.loadMigrate` assumes `project.loadCompile` has already been run [at line 51](https://github.com/trufflesuite/truffle/blob/develop/packages/db/src/project/loadMigrate/index.ts#L51). 

`loadCompile` runs as part of the `workflow-compile` [workflow ](https://github.com/trufflesuite/truffle/blob/develop/packages/workflow-compile/index.js#L126).

When `loadMigrate` is called [in `Migration.js`](https://github.com/trufflesuite/truffle/blob/develop/packages/migrate/Migration.js#L125) to attach the network information to the project in the db, the contract data from `loadMigrate`/`workflow-compile` is not present so it fails [at extracting & attaching that information to the data](https://github.com/trufflesuite/truffle/blob/develop/packages/db/src/project/loadMigrate/contracts.ts#L42).

But wait, there's more...

`workflow-compile` still runs, however, the contracts array is empty, so [length is zero.](https://github.com/trufflesuite/truffle/blob/develop/packages/workflow-compile/index.js#L112). This is because in the `compile-solidity` package the [`necessary` function ](https://github.com/trufflesuite/truffle/blob/develop/packages/compile-solidity/src/index.js#L104) from Profile returns `[]` back from the [`updated`](https://github.com/trufflesuite/truffle/blob/develop/packages/profiler/src/index.ts#L25) function. I.e. no code change detected, the compiler does not run.


Trace:
1. [`Migrate` calls `WorkflowCompile.compileAndSave`](https://github.com/trufflesuite/truffle/blob/develop/packages/core/lib/commands/migrate/run.js#L20)
2. [`compileAndSave` calls `compile`.](https://github.com/trufflesuite/truffle/blob/develop/packages/workflow-compile/index.js#L99)
3.[ `compile` tries each compiler](https://github.com/trufflesuite/truffle/blob/develop/packages/workflow-compile/index.js#L35),
4. `compileAll` = `false`. compileMethod is `require("@truffle/compile-solidity").Compile`
5. `compileMethod` calls [`necessary`](https://github.com/trufflesuite/truffle/blob/develop/packages/workflow-compile/index.js#L33)
6. [`necessary` requests paths from the `Profiler.updated` method](https://github.com/trufflesuite/truffle/blob/develop/packages/compile-solidity/src/index.js#L104)
7. [`Profiler.updated` determines that no files have been updated.](https://github.com/trufflesuite/truffle/blob/develop/packages/profiler/src/index.ts#L40)
8. `compileMethod.necessary` calls [`Compile.sourcesWithDependencies`](https://github.com/trufflesuite/truffle/blob/develop/packages/compile-solidity/src/index.js#L106) with the `paths` parameter = `[]`
9. `WorkflowCompile` calls[ `save` ](https://github.com/trufflesuite/truffle/blob/develop/packages/workflow-compile/index.js#L101)and passes empty values for `{ contracts, sources, compilations}`
10. Although the db is configured, there is no contract data [(`contracts.length === 0`)](https://github.com/trufflesuite/truffle/blob/develop/packages/workflow-compile/index.js#L112) to save

Suggestions:

This is a very deep trace and obviously touches on code that is working as intended for truffle's existing developer workflow. In it's current state, truffleDb is implemented as a 'co-witness' to the data. I see two solutions:

1. Easiest, try/catch the `loadMigrate` function. If the user flips on TruffleDB in the capacity described in this issue we know the exact reason why the error is thrown. Provide an error message to run `compile` with `db:enabled` - NOTE: Handling this error will still allow the migration to complete, as it has already run and TruffleDB is only adding the network data to the contracts it assumed were saved in `load-compile`. If we went ahead and saved the migration data without the `loadCompile` data we hit a possible referential integrity issue where the contract may not be saved, but the migration information exist in the db. 

2. More thorough, but possible side effects. The `Profile.updated` function is unaware of TruffleDB, or that the user could switch TruffleDB on in between the `compile` and `migrate` commands. Thus, although the files exist on disk, it is unaware the data does not exist in the database. At this time, I do not support this path based on the previous feedback that was already covered and think it would become unnecessary tech debt/maintenance and that there may be easier ways to achieve integration into the existing truffle workflow. 


I am happy to hop on a call and share the trace and discuss options. This is not a 'bug' as much as it is an integration issue. The existing truffle workflow is working as intended. Personally, I would prefer to address the way truffledb is implemented as well as how it integrates into the truffle workflow before making changes in working parts of truffle that will need to maintain their backwards compatibility. 

Tagging some folks to get extra eyes on the issue as a general sanity check and for feedback as my experience with the codebase is still limited and there are likely other, better ways to solve the issue. @fainashalts  @cds-amal @eggplantzzz

**Comment by eggplantzzz** [comment]
Yeah this will happen because `Profiler.updated` compares the last edits of both the source files and the built artifacts to determine whether there have been updates since the last compilation. It feels like the proper solution is to update that method to check with the saved db data to see if it is up-to-date when db is enabled. Offhand I'm not certain what that would entail, it might make it necessary for that method to take some sort of parameter to let it know how to check for updated source material. Also happy to chat about this if you want to!
